examples/platform/silabs/rgb_led/RGBLEDWidget.cpp

```diff
@@ -0,0 +1,353 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2019 Google LLC.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "RGBLEDWidget.h"
+#include <math.h>
+
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/silabs/platformAbstraction/SilabsPlatform.h>
+
+/*Delay for PWM simulation*/
+#ifndef TICK_DELAY
+#define TICK_DELAY 2
+#endif
+
+using namespace ::chip::System;
+using namespace chip::DeviceLayer::Silabs;
+
+using RgbColor_t = RGBLEDWidget::RgbColor_t;
+using HsvColor_t = RGBLEDWidget::HsvColor_t;
+
+void RGBLEDWidget::SetColor(uint8_t red, uint8_t green, uint8_t blue)
+{
+    if (GetPlatform().GetRGBLedState(GetLED()))
+    {
+        ChipLogProgress(Zcl, "SetColor : %u|%u|%u", red, green, blue);
+        GetPlatform().SetLedColor(GetLED(), red, green, blue);
+    }
+}
+
+void RGBLEDWidget::GetColor(uint16_t & r, uint16_t & g, uint16_t & b)
+{
+    uint16_t rawR, rawG, rawB;
+    GetPlatform().GetLedColor(GetLED(), rawR, rawG, rawB);
+
+    // Scale down PWM values by TICK_DELAY to bring them back to the expected range
+    // Clamp to 8-bit range (0-255)
+    r = static_cast<uint16_t>((rawR / TICK_DELAY) & 0xFF);
+    g = static_cast<uint16_t>((rawG / TICK_DELAY) & 0xFF);
+    b = static_cast<uint16_t>((rawB / TICK_DELAY) & 0xFF);
+
+    ChipLogProgress(Zcl, "RGB Values: R=%u G=%u B=%u", r, g, b);
+}
+
+void RGBLEDWidget::SetColorFromHSV(uint8_t hue, uint8_t saturation)
+{
+    RGBLEDWidget::HsvColor_t hsv;
+    hsv.h = hue;
+    hsv.s = saturation;
+    hsv.v = GetLevel();
+    ChipLogProgress(Zcl, "SetColorFromHSV : %u|%u", hsv.h, hsv.s);
+    RGBLEDWidget::RgbColor_t rgb = RGBLEDWidget::HsvToRgb(hsv);
+    SetColor(rgb.r, rgb.g, rgb.b);
+}
+
+void RGBLEDWidget::SetColorFromXY(uint16_t currentX, uint16_t currentY)
+{
+    ChipLogProgress(Zcl, "SetColorFromXY: %u|%u", currentX, currentY);
+    RGBLEDWidget::RgbColor_t rgb = RGBLEDWidget::XYToRgb(GetLevel(), currentX, currentY);
+    SetColor(rgb.r, rgb.g, rgb.b);
+}
+
+void RGBLEDWidget::SetColorFromCT(uint16_t ctMireds)
+{
+    RGBLEDWidget::RgbColor_t rgb = RGBLEDWidget::CTToRgb(ctMireds);
+    ChipLogProgress(Zcl, "SetColorFromCT: %u", ctMireds);
+    SetColor(rgb.r, rgb.g, rgb.b);
+}
+
+// define a clamp macro to substitute the std::clamp macro which is available from C++17 onwards
```

Comment:
we are in C++17. Use std::clamp

Reply:
Replaced with std::clamp. Thanks for your comment.